clients/src/main/java/org/apache/kafka/clients/producer/internals/RecordAccumulator.java

```diff
@@ -212,24 +214,43 @@ public RecordAppendResult append(TopicPartition tp,
      * Abort the batches that have been sitting in RecordAccumulator for more than the configured requestTimeout
      * due to metadata being unavailable
      */
-    public List<RecordBatch> abortExpiredBatches(int requestTimeout, Cluster cluster, long now) {
+    public List<RecordBatch> abortExpiredBatches(int requestTimeout,
+                                                 Cluster cluster,
+                                                 KafkaClient client,
+                                                 long now) {
         List<RecordBatch> expiredBatches = new ArrayList<RecordBatch>();
         int count = 0;
         for (Map.Entry<TopicPartition, Deque<RecordBatch>> entry : this.batches.entrySet()) {
             Deque<RecordBatch> dq = entry.getValue();
-            synchronized (dq) {
-                // iterate over the batches and expire them if they have stayed in accumulator for more than requestTimeOut
-                Iterator<RecordBatch> batchIterator = dq.iterator();
-                while (batchIterator.hasNext()) {
-                    RecordBatch batch = batchIterator.next();
-                    // check if the batch is expired
-                    if (batch.maybeExpire(requestTimeout, now, this.lingerMs)) {
-                        expiredBatches.add(batch);
-                        count++;
-                        batchIterator.remove();
-                        deallocate(batch);
-                    } else {
-                        if (!batch.inRetry()) {
+            // We will check if the batch should be expired if one of the following is true:
+            // 1. The leader is unknown.
+            // 2. The leader broker is disconnected.
+            Node leader = cluster.leaderFor(entry.getKey());
+            boolean leaderNotConnected = true;
+            try {
+                leaderNotConnected = client.connectionFailed(leader);
+            } catch (IllegalArgumentException e) {
+                // This means the client does not know the leader node. So it is not connected.
+            }
+
+            if (leader == null || leaderNotConnected) {
```

Comment:
@junrao Not sure if I missed something. In order to time out Batch 2 at T + 31 secs, the NetworkClients has to be disconnected from the leader broker at that point. When NetworkClient disconnects from a broker, it will clear the in-flight request in processDisconnection(). And in Sender.handleProduceResponse(), the sender will re-enque the failed batch when the ClientResponse.disconnected=true if retries are not exhausted. 

That means if NetworkClient.connectionFailed() returns true, there is no in flight request to that broker.

So in your example, at T + 31 sec (before Batch 1 hit request timeout hit at T + 10 + 30 secs), if the NetworkClient disconnects from the broker, Batch 1 will be re-enqued right after the NetworkClient update the node status to DISCONNECTED. So in the next poll() when the partition sees its leader is disconnected, batch 1 should have already been re-enqued. Since batch 1 has just been re-enqued, it won't be timed out immediately. And because we break out of the expiration checking while loop once we see a batch that should not be timed out. Batch 2 will not be timed out either.

You are right that from timeout ordering perspective, it is fine to use the muted partitions. And it is also simpler. I was just trying to see if we can make the producer more robust to the server failure because that was also the original motivation of adding the request timeout.